flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/SearchOperatorGen.scala

```diff
@@ -97,6 +97,9 @@ object SearchOperatorGen {
            |if (!${needle.nullTerm}) {
            |  $resultTerm = $negation$setTerm.contains(${needle.resultTerm});
            |  $nullTerm = !$resultTerm && $setTerm.containsNull();
+           |} else if (${sarg.nullAs == RexUnknownAs.TRUE}) {
+           |  $resultTerm = $negation$setTerm.containsNull();
+           |  $nullTerm = !$resultTerm && $setTerm.containsNull();
```

Comment:
This is incorrect. Try `SELECT * FROM myTable WHERE a NOT IN (1, 2) or a IS NULL`.

`sarg.nullAs` will be `RexUnknownAs.TRUE`, however `negation` will be `!` so if `a` is `NULL` this expression will be false. This is obviously incorrect.

I guess `sarg.nullAs` defines what to return for `NULL`s, regardless of `sarg.isComplementedPoints`. As currently I'm mainly working on Flink Table Store I can't guarantee this guess is correct. You need to check for yourself.

I suggest you try different combinations of `(NOT) IN` and `IS (NOT) NULL`. Also try to add more test cases in `CalcITCase` to verify your changes for different expressions.

Reply:
thanks for your feedback, i was able to reproduce the issue
will have a look and  add more tests for that